### ISSUE TITLE
[2020 CVPR] Designing Network Design Spaces

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#50]
기존의 NAS의 단점으로는 single network instance를 뱉지 network design principle을 알 순 없음

논문은 network design space를 알아내어 network design principle을 알아내자는 것임

AnyNet이라는 거의 제약 없는 네트워크에서부터 시작하여 많은 모델들의 성능을 측정하며 사람이 제약을 걸어 RegNet 시리즈들을 차차 만들어감 (human-in-the-loop methodology)

## 1. Tools for Design Space Design

Design space의 quality를 측정하기 위해 design space에 해당하는 모델들을 sample하여 학습한 후, 모델들의 error distribution을 이용

즉 distribution간의 비교를 통해 보다 더 robust하고 informative한 design space를 찾고, 2 design space가 있을 때 가장 좋은 모델들 간에 비교할 수 있음

Distribution을 구하기 위해 design space에서 n개의 모델을 sample하여 train함

Efficiency를 위해 low-compute, low-epoch training regime로 수행 (ImageNet 224 input resolution, 10 epoch만 수행)

Error EDF (empirical distribution function)를 이용하여 design space의 quality를 분석

<img src='https://user-images.githubusercontent.com/49643709/152524963-be5e9666-680a-45cb-ac27-6c30ad3f772e.png' width=80% height=80%>

그리고 학습된 모델들의 network property와 error를 plot한 후, empirical bootstrap 방법을 이용하여 design space의 insight를 얻음

<img src='https://user-images.githubusercontent.com/49643709/152525423-08111c1e-b5cb-47bd-b5ed-5eca1265b9e8.png' width=80% height=80%>

즉 해당 논문의 process는 다음 과정을 반복하는 것임

1. Design space에 해당하는 n개의 모델들을 학습하여 model distribution 생성
2. Error EDF를 이용하여 design space quality 측정
3. Design space의 다양한 property를 plot한 후 empirical bootstrap을 이용하여 insight를 얻음
4. 이러한 insight를 기반으로 design space에 제약을 추가

### 2.1. The AnyNet Design Space
<img src='https://user-images.githubusercontent.com/49643709/152556914-908f8f06-3845-4366-8297-c99af4716f55.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152557027-5b1ad2a2-ab07-47df-ae89-a3d0027a6583.png' width=80% height=80%>

- 제약이 없는 초기 버전인 AnyNet 모델의 설계는 다음과 같음
    - 네트워크는 stem, body, head로 구성되어있으며 body 파트만 고려 (stem, head 고정)
    - body는 4 stage로 구성되며, stage가 증가하면서 resolution이 점진적으로 감소
    - 각 stage는 동일한 block을 사용
    - network property : block num, width, bottleneck ratio, group width
    - group convolution이 있는 standard residual bottleneck block을 default로 세팅 (ResNeXt 모듈이라 생각하면 됨)

AnyNetX design space의 자유도는 4 (stage) * 4 (parameter)로 16이고 각 parameter별 범위는 다음과 같음

<img src='https://user-images.githubusercontent.com/49643709/152556242-9b327538-660d-463a-9e3a-5e3a0b2f6aa5.png' width=80% height=80%>

### 2.2. AnyNet Experiments
<img src='https://user-images.githubusercontent.com/49643709/152560683-bfc6d0bc-8242-4277-b2f8-79219bdd0b23.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152560736-696e6e93-c3f2-412d-a773-52e666cb15ae.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152560768-71515ba3-5993-4db7-b391-948339afd093.png' width=80% height=80%>

|AnyNet|Constraints|Details|
|:-:|:-:|:-:|
|AnyNetX_A| - | Initial AnyNetX design space |
|AnyNetX_B| AnyNetX_A + Share bottleneck ratio for all stages | |
|AnyNetX_C| AnyNetX_B + Share group width for all stages | group width > 1 is best |
|AnyNetX_D| AnyNetX_C + Increasing widths | w_i+1 >= w_i |
|AnyNetX_E| AnyNetX_D + Increasing depths | d_i+1 >= d_i, not necessary in the last stage |

### 2.3. The RegNet Design Space
<img src='https://user-images.githubusercontent.com/49643709/152561342-bcdb77e6-c96e-4be1-a80b-2aaa430728e1.png' width=80% height=80%>

AnyNetX_E design space에서 best 20 models에 대해 plot한 결과 width와 block간의 선형 식을 유추할 수 있었음

하지만 이 선형식을 그대로 사용할 수 없기에, quantize하여 piecewise constant function으로 만듬

### 2.4. Design Space Summary
<img src='https://user-images.githubusercontent.com/49643709/152563476-6228e637-73c1-4c09-9586-0729ef091f3d.png' width=80% height=80%>

### 2.5. Design Space Generalization
<img src='https://user-images.githubusercontent.com/49643709/152563797-83c004b0-187c-46ca-ae58-ae1f4c1a7ae4.png' width=80% height=80%>

AnyNetX_A, AnyNetX_E, RegNetX design space를 다양한 조건에서 학습을 진행했는데 모든 경우에서 RegNetX > AnyNetX_E > AnyNetX_A 순서로 일관된 결과를 보여줌
(higher flops, higher epochs, with 5-stage networks, with various block types)

즉 signs of overfitting을 찾아볼 수 없었으며, RegNetX design space는 새로운 세팅에서도 generalize가 잘 된다는 의미임

### 3.1. RegNet trends
<img src='https://user-images.githubusercontent.com/49643709/152632755-31ec255a-c471-484c-9760-5d7650cac12b.png' width=80% height=80%>

- RegNetX design space에서 100개의 model을 sample하여 25 epoch 동안 학습한 결과, best models 사이에 공통적인 패턴 존재
    - optimal depth ~20 blocks (60 layers)
    - bottleneck ratio 1.0
    - width multiplier ~2.5

### 3.2. Complexity analysis
<img src='https://user-images.githubusercontent.com/49643709/152632816-89f79dc4-f845-4fae-9621-5d5c8ca8baba.png' width=80% height=80%>

Flops, parameters과 더불어 activations를 이용하여 complexity 분석

Best model일 수록 complexity 증가

### 3.3. RegNetX constrained
<img src='https://user-images.githubusercontent.com/49643709/152632939-12f2e347-66c3-41a9-8aff-b3f7d1bc73d4.png' width=80% height=80%>

- RegNet trend에서 밝혀진 결과들을 이용하여 다시 제약 조건 추가
    - bottleneck ratio = 1
    - depth <= 40
    - width >= 2
    - limit parameters and activations

실험 결과 제약 조건을 추가한 버전이 더 우수하며, depth limit을 12 <= d <= 28로 세팅

### 3.4. Alternate design choices

- Inverted bottleneck : 효과 X (EDF degrade)
- Depthwise conv : 효과 X (b = 1, g >= 1인 design space보다 성능이 낮음)
- Scaling resolution : 224 * 224 fixed resolution이 더 효과 있음 (even at higher flops)

### 3.5. SE
<img src='https://user-images.githubusercontent.com/49643709/152633124-9c3c9931-a2bf-4bcb-8b5a-1828125dd808.png' width=80% height=80%>

RegNetX에 Squeeze-and-Excitation (SE) op을 추가한 버전을 RegNetY로 명칭

RegNetY는 성능 gain이 있었음 (즉 SE의 효과가 있었음)

### 4.1. Pattern
<img src='https://user-images.githubusercontent.com/49643709/152633172-acd390ce-7911-40bc-8dfb-dd3d58aff742.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633184-9123ec86-6d87-4e08-94b9-10666047be50.png' width=80% height=80%>

- Higher flop model의 경우 3 stage에서 block 수가 많고 4 stage에서는 block 수가 적음
- Complexity가 증가하면 group width가 증가
- Depth는 saturate

### 4.2. State-of-the-Art Comparison: Mobile Regime
<img src='https://user-images.githubusercontent.com/49643709/152633289-dedd9a2b-bed6-4f00-9450-be92a6c4147a.png' width=80% height=80%>

기존의 모바일에 최적화된 모델들은 다양한 방법들을 이용하여 학습했는데, 이런 테크닉 없이도 성능이 잘 나옴
(while most mobile networks use longer schedules with various enhancements, such as deep supervision, Cutout, DropPath, AutoAugment, ...)

### 4.3. Standard Baselines Comparison : ResNe(X)t
<img src='https://user-images.githubusercontent.com/49643709/152633372-6a41db95-66b2-4d14-ac44-d18dfb097906.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633380-e175786f-7c37-4dc4-871a-be206bef10d9.png' width=80% height=80%>

기존 baseline보다 성능이 좋음

### 4.4. State-of-the-Art Comparison : Full Regime
<img src='https://user-images.githubusercontent.com/49643709/152633425-e0ab8c2d-9bda-4c38-a381-59b2330923d4.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633428-76b38c75-8138-4108-8e9a-79052314bc3f.png' width=80% height=80%>

Low flops에서는 EfficientNet > RegNetY

Intermediate flops에서는 RegNetY > EfficientNet

Higher flops에서는 RegNetX, RegNetY > EfficientNet

그리고 RegNet의 경우 activation 수가 적어 EfficientNet보다 train, inference time이 훨씬 작아서 유리함

### 5.1. Appendix A : Test Set Evaluation
<img src='https://user-images.githubusercontent.com/49643709/152633516-bd8d5dfe-7cba-4cb5-8266-9ad545fddb7d.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633522-611cb133-820f-4b47-844c-02e520658b6d.png' width=80% height=80%>

### 5.2. Appendix B : Additional Ablations
<img src='https://user-images.githubusercontent.com/49643709/152633593-ec4d4b0b-627f-4379-a571-0cdeda309b84.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633731-a69f4c50-b0f6-4035-9b9e-ee431f24be98.png' width=80% height=80%>

#### 5.2.1. Fixed Depth

Depth의 경우 saturation이 되는 현상이 일어났기에, depth = 20으로 fix하여 제약 조건을 걸어 실험 수행

Fixed-depth networks로도 variable depth networks의 performance를 match함 (for all flop regimes)

즉 depth를 fix해도 된다는 말

#### 5.2.2. Fewer stages

Last stage의 block이 매우 적기에, 3 stage network로 실험을 진행했는데 performance degrade

Stem 혹은 head를 고치면 3 stage network도 충분히 잘할 수 있을 것 같으나 future work로 남겨둠

#### 5.2.3. Inverted Bottleneck

Inverted bottleneck을 사용하면 performance degrade (b < 1)

#### 5.2.4. Swish vs ReLU

Low flops에서는 Swish > ReLU

Higher flops에서는 ReLU > Swish

Group width가 1로 고정일 경우 Swish > ReLU (g=1, depthwise conv)

### 5.3. Appendix C : Optimization Settings
<img src='https://user-images.githubusercontent.com/49643709/152633825-afec56e9-20fb-4f1b-8ccf-0800403a45b1.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633834-ff6d01c1-7031-4891-9a26-54f1732cb3d2.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152633840-b3a32c3d-bca0-4941-bfe7-62c67b39b1c4.png' width=80% height=80%>

- Learning rate = 0.1
- Weight decay = 0.00005
- Adjust learning rate using the linear scaling rule and apply 5 epoch gradual warmup

### 5.4. Appendix D : Implementation Details
<img src='https://user-images.githubusercontent.com/49643709/152634111-aa44b476-befd-4447-873b-9703d71f2195.png' width=80% height=80%>

#### 5.4.1. Group width compatibility

Sample된 width가 group width로 나뉘어지지 않을 수 있는데, 이를 해결하기 위해 w가 나뉘어질 수 있게 조절해줌
(g > w인 경우 g = w로 바꿔줌)

Bottleneck이 있는 모델들의 경우에도 비슷하게 진행

#### 5.4.2. Group width ranges

Higher flops에서 group width가 증가하는 추세가 있기에, higher compute regimes에서는 group width range를 조절해줌
(16 <= g <= 64 instead of g <= 32)

#### 5.4.3. Block types

RegNet design space가 다양한 block type에도 적용됨을 보여주기 위해 실험 진행

1. R block : same as X block except without groups
2. V block : a basic block with only a single 3 * 3 conv
3. VR block : same as V block plus residual connections

각 block마다 좋은 parameter value가 다르지만 RegNet design space로도 충분함

#### 5.4.4. Y block details

SE op는 X block의 3 * 3 conv 이후에 적용하며 SE reduction ratio는 0.25로 설정 (1/4)
